[material/menu_anchor.dart] Remove _MenuAnchorState from parent when disposed.

davidhicks980 opened the pull request:
`_MenuAnchorState` now removes itself from its parent regardless of whether `_MenuAnchorState._isOpen` is true.

Before, _MenuAnchorState would only detach itself when `_MenuAnchorState._isOpen == true`. This led to _MenuAnchorState being retained until an anchor's parent was disposed.

Before:

https://github.com/flutter/flutter/assets/59215665/48f1617a-a3d0-49ab-a767-f41e3a30ea41

After:

https://github.com/flutter/flutter/assets/59215665/31312bb6-49ec-4083-b8de-e9ae2a42a8b3

Fixes https://github.com/flutter/flutter/issues/149584

I was a bit unsure what the best way of testing this change would be, and wanted feedback from the Flutter team. Is there a way for us to view at the instance count of `_MenuAnchorState`, or expose `_MenuAnchorState._anchorChildren.length` for testing?

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

davidhicks980 committed:
Update menu_anchor.dart

Removed self from parent regardless of whether menu is open.

Piinks commented:
@davidhicks980 thanks for sending the PR! Would you mind adding a test?

polina-c commented:
> @polina-c I am surprised the leak tracker did not already catch this. Do you know how we might be able to capture it in a test?

LeakTracker is scoped to catch not disposed leaks at the moment. It does not catch not GCed ones.
It is separate big effort to clean up Flutter Framework from notGCed leaks. And it is not part of leak tracker MVP.
This PR fixes notGCed leak.

Does it help?

Piinks commented:
Ah ok, thank you for calrifying.

Piinks commented:
I was looking at the original issue,

> #### Expected results
>There should only be one _MenuAnchorState for each menu anchor
>
> #### Actual results
>Detached _MenuAnchorStates are being retained

Maybe to test @davidhicks980 we could just access `currentState` without needing to expose _MenuAnchorState and just check it is not the same state after opening/closing/opening the menu?

davidhicks980 commented:
@Piinks So I may be wrong about this, but I think the currentState would still correctly report that it has changed since the state attached to the visible MenuAnchor is changing, but the parent _MenuAnchorState would still be holding on to previous instances of the state as well.

To verify, I wrote this test with the unfixed menu and it did pass, but my test could also be flawed:

```dart
testWidgets('Nested anchor state should dispose after a menu is closed', (WidgetTester tester) async {
      dynamic oldState;
      GlobalKey stateKey = GlobalKey();
      await tester.pumpWidget(MaterialApp(
        home: Material(
          child: Center(
            child: MenuAnchor(
               controller: controller,
                builder: (context, controller, child) {
                  return TextButton(
                      onPressed: () {
                        if (controller.isOpen) {
                          controller.close();
                        } else {
                          controller.open();
                        }
                      },
                      child: const Text('Toggle Menu'));
                },
                menuChildren: [
                  SubmenuButton(
                    key: stateKey,
                    menuChildren: [
                      MenuItemButton(
                        onPressed: () {},
                        child: const Text('Item 1'),
                      ),
                      MenuItemButton(
                        onPressed: () {},
                        child: const Text('Item 2'),
                      ),
                    ],
                    child: const Text('Item 2'),
                  ),
                  SubmenuButton(
                    menuChildren: [
                      MenuItemButton(
                        onPressed: () {},
                        child: const Text('Item 1'),
                      ),
                      MenuItemButton(
                        onPressed: () {},
                        child: const Text('Item 2'),
                      ),
                    ],
                    child: const Text('Item 2'),
                  ),
                ]),
          ),
        ),
      ));
      controller.open();
      await tester.pump();
      oldState = stateKey.currentState;
      controller.close();
      await tester.pump();
      controller.open();
      await tester.pump();
      expect(stateKey.currentState, isNot(oldState));
    });
```

davidhicks980 committed:
Added forced GC test.

Added a test that pairs forces garbage collection on a weakly-referenced _MenuAnchorState to check whether it has been dereferenced when its parent is closed.

davidhicks980 committed:
Renamed test

davidhicks980 committed:
Reworded again to remove posessive words

davidhicks980 committed:
Merge branch 'flutter:master' into patch-3

davidhicks980 commented:
Okay, assuming checks pass, and forceGC works on CI, this should be good-to-go! I added that forces garbage collection on a weakly-referenced _MenuAnchorState, and it appears to correctly identify if a child state is retained by its parent.

davidhicks980 committed:
Added test skip on kIsWeb 

ForceGC doesn't work on web

davidhicks980 committed:
Added leak_tracker.dart dependency

davidhicks980 committed:
Fixed CI issues

davidhicks980 committed:
Added multiple GC runs

davidhicks980 committed:
Merge branch 'master' into patch-3

Piinks commented on packages/flutter/test/material/menu_anchor_test.dart:
```diff
@@ -3803,6 +3804,51 @@ void main() {
     );
   });
+
+  // Fixes https://github.com/flutter/flutter/pull/149586
```
```suggestion
```

Piinks commented on packages/flutter/test/material/menu_anchor_test.dart:
```diff
@@ -3803,6 +3804,51 @@ void main() {
+
+  // Fixes https://github.com/flutter/flutter/pull/149586
+  testWidgets('Garbage collector destroys child _MenuAnchorState after parent is closed', (WidgetTester tester) async {
+      await tester.pumpWidget(
```
```suggestion
      // Regression test for https://github.com/flutter/flutter/pull/149586
      await tester.pumpWidget(
```

Piinks commented on packages/flutter/lib/src/material/menu_anchor.dart:
```diff
@@ -328,8 +328,11 @@ class _MenuAnchorState extends State<MenuAnchor> {
-      _parent?._removeChild(this);
     }
+
+    // Fix for https://github.com/flutter/flutter/issues/149584
```
```suggestion
```

Piinks reviewed (commented):
Thanks for adding a test - this LGTM with just a few nits

Piinks committed:
Apply suggestions from code review

justinmc commented on packages/flutter/test/material/menu_anchor_test.dart:
```diff
@@ -3803,6 +3804,51 @@ void main() {
   });
+
+  testWidgets('Garbage collector destroys child _MenuAnchorState after parent is closed', (WidgetTester tester) async {
+      // Regression test for https://github.com/flutter/flutter/pull/149586
```
Should you link to the issue? https://github.com/flutter/flutter/issues/149584

Piinks replied:
```suggestion
      // Regression test for https://github.com/flutter/flutter/issues/149584
```

justinmc reviewed (approved):
Thank you for fixing this! Good catch. LGTM :+1: 

FYI @gspencergoog because MenuAnchor.

Piinks committed:
Update packages/flutter/test/material/menu_anchor_test.dart
